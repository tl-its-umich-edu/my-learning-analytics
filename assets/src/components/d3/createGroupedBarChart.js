import * as d3 from 'd3'
import { adjustViewport } from '../../util/chart'
import { margin } from '../../constants/chartConstants'
import { getValues, flatten } from '../../util/array'

function createGroupedBarChart ({ data, width, height, tip, el }) {
  const [aWidth, aHeight] = adjustViewport(width, height, margin)

  const x0 = d3.scaleBand()
    .domain(data.map(d => d.label))
    .range([margin.left, aWidth - margin.right])
    .padding(0.1)

  const keys = Object.keys(data[0].data)

  const x1 = d3.scaleBand()
    .domain(keys)
    .rangeRound([0, x0.bandwidth()])
    .padding(0.5)

  const y = d3.scaleLinear()
    .domain([0, d3.max(flatten(data.map(x => getValues(x.data))))]).nice()
    .range([aHeight - margin.bottom, margin.top])

  const colours = d3.scaleOrdinal()
    .range(['#98abc5', '#8a89a6', '#7b6888', '#6b486b', '#a05d56', '#d0743c', '#ff8c00'])

  const svg = d3.select(el).append('svg')
    .attr('width', aWidth)
    .attr('height', aHeight)

  const g = svg.append('g')
    .selectAll('g')
    .data(data).enter()
    .append('g')
    .attr('transform', d => `translate(${x0(d.label)}, 0)`)

  const bar = g.selectAll('rect')
    .data(d => keys.map(key => ({ key, data: d.data[key] })))
    .enter().append('rect')
    .attr('x', d => x1(d.key))
    .attr('y', d => y(d.data))
    .attr('width', x1.bandwidth())
    .attr('height', d => aHeight - margin.bottom - y(d.data))
    .attr('fill', d => colours(d.key))

  svg.append('g')
    .attr('transform', `translate(0,${aHeight - margin.bottom})`)
    .call(d3.axisBottom(x0))

  svg.append('g')
    .attr('transform', `translate(${margin.left},0)`)
    .call(d3.axisLeft(y).ticks(null, 's'))
    .append('text')
    .attr('x', 2)
    .attr('y', y(y.ticks().pop()) + 0.5)
    .attr('dy', '0.32em')
    .attr('fill', '#000')
    .attr('font-weight', 'bold')
    .attr('text-anchor', 'start')
    .text('Population')

  if (tip) {
    svg.call(tip)
    bar
      .on('mouseover', tip.show)
      .on('mouseout', tip.hide)
  }
}

export default createGroupedBarChart
