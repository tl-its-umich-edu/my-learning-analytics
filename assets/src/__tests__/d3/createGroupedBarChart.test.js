/* global describe, test, expect */
import createGroupedBarChart from '../../components/d3/createGroupedBarChart'
import { groupedBarChartData } from '../testData/d3TestData'

describe('createGroupedBarChart', () => {
  test('should build a grouped bar chart', () => {
    const div = document.createElement('div')
    createGroupedBarChart({ data: groupedBarChartData, width: 1000, height: 500, el: div })
    expect(div).toMatchSnapshot()
  })
})
