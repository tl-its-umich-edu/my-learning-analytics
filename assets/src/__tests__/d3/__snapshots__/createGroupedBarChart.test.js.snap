// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createGroupedBarChart should build a grouped bar chart 1`] = `
<div>
  <svg
    height="450"
    width="940"
  >
    <g>
      <g
        transform="translate(57.25490196078431, 0)"
      >
        <rect
          fill="#98abc5"
          height="98.35123636363636"
          width="10"
          x="13"
          y="321.64876363636364"
        />
        <rect
          fill="#8a89a6"
          height="163.6323636363636"
          width="10"
          x="33"
          y="256.3676363636364"
        />
        <rect
          fill="#7b6888"
          height="78.54476363636365"
          width="10"
          x="53"
          y="341.45523636363635"
        />
        <rect
          fill="#6b486b"
          height="140.13774545454544"
          width="10"
          x="73"
          y="279.86225454545456"
        />
        <rect
          fill="#a05d56"
          height="385.6185454545455"
          width="10"
          x="93"
          y="34.38145454545452"
        />
        <rect
          fill="#d0743c"
          height="320.70334545454546"
          width="10"
          x="113"
          y="99.29665454545454"
        />
        <rect
          fill="#ff8c00"
          height="149.61803636363635"
          width="10"
          x="133"
          y="270.38196363636365"
        />
      </g>
      <g
        transform="translate(229.80392156862743, 0)"
      >
        <rect
          fill="#98abc5"
          height="73.72025454545451"
          width="10"
          x="13"
          y="346.2797454545455"
        />
        <rect
          fill="#8a89a6"
          height="119.19803636363639"
          width="10"
          x="33"
          y="300.8019636363636"
        />
        <rect
          fill="#7b6888"
          height="51.65519999999998"
          width="10"
          x="53"
          y="368.3448"
        />
        <rect
          fill="#6b486b"
          height="89.26258181818184"
          width="10"
          x="73"
          y="330.73741818181816"
        />
        <rect
          fill="#a05d56"
          height="255.19021818181818"
          width="10"
          x="93"
          y="164.80978181818182"
        />
        <rect
          fill="#d0743c"
          height="205.6919272727273"
          width="10"
          x="113"
          y="214.3080727272727"
        />
        <rect
          fill="#ff8c00"
          height="89.89901818181818"
          width="10"
          x="133"
          y="330.1009818181818"
        />
      </g>
      <g
        transform="translate(402.35294117647055, 0)"
      >
        <rect
          fill="#98abc5"
          height="41.47330909090908"
          width="10"
          x="13"
          y="378.5266909090909"
        />
        <rect
          fill="#8a89a6"
          height="56.687963636363634"
          width="10"
          x="33"
          y="363.31203636363637"
        />
        <rect
          fill="#7b6888"
          height="33.638545454545465"
          width="10"
          x="53"
          y="386.36145454545454"
        />
        <rect
          fill="#6b486b"
          height="89.26258181818184"
          width="10"
          x="73"
          y="330.73741818181816"
        />
        <rect
          fill="#a05d56"
          height="173.89523636363637"
          width="10"
          x="93"
          y="246.10476363636363"
        />
        <rect
          fill="#d0743c"
          height="172.61294545454547"
          width="10"
          x="113"
          y="247.38705454545453"
        />
        <rect
          fill="#ff8c00"
          height="115.9198909090909"
          width="10"
          x="133"
          y="304.0801090909091"
        />
      </g>
      <g
        transform="translate(574.9019607843137, 0)"
      >
        <rect
          fill="#98abc5"
          height="32.522472727272714"
          width="10"
          x="13"
          y="387.4775272727273"
        />
        <rect
          fill="#8a89a6"
          height="56.687963636363634"
          width="10"
          x="33"
          y="363.31203636363637"
        />
        <rect
          fill="#7b6888"
          height="26.399018181818178"
          width="10"
          x="53"
          y="393.6009818181818"
        />
        <rect
          fill="#6b486b"
          height="47.69014545454547"
          width="10"
          x="73"
          y="372.3098545454545"
        />
        <rect
          fill="#a05d56"
          height="130.77610909090907"
          width="10"
          x="93"
          y="289.2238909090909"
        />
        <rect
          fill="#d0743c"
          height="117.78810909090907"
          width="10"
          x="113"
          y="302.2118909090909"
        />
        <rect
          fill="#ff8c00"
          height="57.28392727272728"
          width="10"
          x="133"
          y="362.7160727272727"
        />
      </g>
      <g
        transform="translate(747.4509803921568, 0)"
      >
        <rect
          fill="#98abc5"
          height="26.8168"
          width="10"
          x="13"
          y="393.1832"
        />
        <rect
          fill="#8a89a6"
          height="48.92149090909089"
          width="10"
          x="33"
          y="371.0785090909091"
        />
        <rect
          fill="#7b6888"
          height="24.698218181818163"
          width="10"
          x="53"
          y="395.30178181818184"
        />
        <rect
          fill="#6b486b"
          height="43.779781818181846"
          width="10"
          x="73"
          y="376.22021818181815"
        />
        <rect
          fill="#a05d56"
          height="114.82760000000002"
          width="10"
          x="93"
          y="305.1724"
        />
        <rect
          fill="#d0743c"
          height="124.14549090909088"
          width="10"
          x="113"
          y="295.8545090909091"
        />
        <rect
          fill="#ff8c00"
          height="69.4753090909091"
          width="10"
          x="133"
          y="350.5246909090909"
        />
      </g>
    </g>
    <g
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="middle"
      transform="translate(0,420)"
    >
      <path
        class="domain"
        d="M40.5,6V0.5H920.5V6"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(134.9019607843137,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          CA
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(307.45098039215685,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          TX
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(479.99999999999994,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          FL
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(652.5490196078431,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          IL
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(825.0980392156862,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          PA
        </text>
      </g>
    </g>
    <g
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="end"
      transform="translate(40,0)"
    >
      <path
        class="domain"
        d="M-6,420.5H0.5V20.5H-6"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(0,420.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          0M
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,384.1363636363636)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          1M
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,347.77272727272725)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          2M
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,311.40909090909093)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          3M
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,275.0454545454545)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          4M
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,238.6818181818182)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          5M
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,202.31818181818184)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          6M
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,165.95454545454547)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          7M
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,129.59090909090907)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          8M
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,93.22727272727269)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          9M
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,56.863636363636374)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          10M
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,20.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          11M
        </text>
      </g>
      <text
        dy="0.32em"
        fill="#000"
        font-weight="bold"
        text-anchor="start"
        x="2"
        y="20.5"
      >
        Population
      </text>
    </g>
  </svg>
</div>
`;
